client/finality-grandpa/src/authorities.rs

```diff
@@ -423,9 +423,21 @@ where
 			fork_tree::FinalizationResult::Changed(change) => {
 				status.changed = true;
 
-				// if we are able to finalize any standard change then we can
-				// discard all pending forced changes (on different forks)
-				self.pending_forced_changes.clear();
+				let pending_forced_changes = std::mem::replace(
+					&mut self.pending_forced_changes,
+					Vec::new(),
+				);
+
+				for change in pending_forced_changes {
+					// we will keep all forced change for any later blocks and that are a
+					// descendent of the finalized block (i.e. they are from this fork).
```

Comment:
I guess the intuition of the previous code was, that if finality gets moving again we don't need to force the change.

But then I guess what happened is the runtime and node-side get out of sync?

Reply:
Yep, that was the intuition, and indeed the issue is that the runtime and the client could go out of sync. I think it's safer to instead assume an invariant that *IF* a forced change is signalled then it must be applied (unless we re-org before reaching it's effective number). I think this assumption is safe provided that forced changes will be triggered manually through a governance process and we provide a delay in blocks long enough to be safe against re-orgs.